### DISCUSSION TITLE
Type narrowing with inner functions

### DISCUSSION THREAD

**Discussion by decorator-factory** [microsoft/pyright#2332] [Ideas]
I was wondering why this says "Never":
```py
def f():
    answer: Optional[int] = None

    def set_answer():
        nonlocal answer
        answer = 42

    yield set_answer
    assert answer is not None
    reveal_type(answer)
```
I read #851 and it made sense.

I was trying to make this work:
```py
class Storage:
    def __init__(self):
        self._files: dict[str, bytes] = {}

    async def find(self, key: str) -> Optional[AsyncIterator[bytes]]:
        content = self._files.get(key)
        if content is None:
            return None

        async def _find() -> AsyncIterator[bytes]:
            yield content
        return _find()
```
Pyright is saying that `content` could be `None`. I found an explanation in #599 and #1731 but found it strange that it behaves like this.

I wondered if TypeScript behaved the same but it didn't: [(TypeScript playground)](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAZgSzAEwGIIDYFMYF4YAU0IATgIYDmWAXDALJkAOAPNCUhQDQxscB8ASjx9CAaywBPWrzAUBtAJJQs5AEbZWUdrJEAfGGACuGDMJgBvAFAwbMUJFgA3MhkM58xclQB0VKAXEJAUtrWwQ4QmdXd1x8QxQsRDAsZGDbdJgSLChDEjADYwwQ9PtoTMc8eHjgKARwACpCISsM2wkELAxkRqi3UJsAX37M7Nz8kkcCYIGgA)

However, if you change `const value` to `let value`, TypeScript does what pyright does. Would it be possible for pyright to have typescript's `const` semantics if it knows that there's only one assignment?

**Comment by erictraut** [comment]
Even if pyright assumed const semantics, it would still need to perform code flow analysis of the outer scope and "guess" where in the code flow the inner function was called. This can't be determined in the general case. Consider the following:

```python
def call_something(func: Callable[[], AsyncIterator[bytes]]) -> AsyncIterator[bytes]:
    return func()

class Storage:
    async def find(self, key: str) -> Optional[AsyncIterator[bytes]]:
        content = self._files.get(key)

        async def _find() -> AsyncIterator[bytes]:
            yield content

        call_something(_find)

        if content is None:
            return None

        return call_something(_find)
```

The workaround I suggest here is to use a new temporary variable whose type is known not to be `None` and can be captured by the inner function.

```python
    async def find(self, key: str) -> Optional[AsyncIterator[bytes]]:
        content = self._files.get(key)
        if content is None:
            return None
        validated_content = content
        
        async def _find() -> AsyncIterator[bytes]:
            yield validated_content
```
